Outline Rust project 4 (#70)
pingcap/talent-plan: rust/projects/parallelism/project.md

In this project you will create a simple key/value server and client that
communicate over a custom protocol. The server will use synchronous networking,
and will respond to multiple requests using increasingly sophisticated
parallel implementations. The in-memory index will become a concurrent
data structure, shared by all threads, and compaction will be done on a
dedicated thread, to reduce latency of individual requests.

This is the final project before we get to the real heavy stuff &mdash; async.
Be afraid...

## Project spec

The cargo project, `kvs`, builds a command-line key-value store client called
`kvs-client`, and a key-value store server called `kvs-server`, both of which in
turn call into a library called `kvs`. The client speaks to the server over
a custom protocol.

The interface to the CLI is the same as in the [previous project]. The
difference this time is in the parallel implementation, which will be described
as we work through it..

The library interface is nearly the same except for two things. First this time
all the `KvsEngine`, `KvStore`, etc. methods take `&self` instead of `&mut
self`, and now it implements `Clone`. This is common with parallel

> datastructions. Why is that? It's not that we're not going to be writing

Reviewer: ```suggestion
data structures. Why is that? It's not that we're not going to be writing
```

immutable code. It _is_ though going to be shared across threads. Why might that
preclude using `&mut self` in the method signatures? If you don't know now,
it'll become obvious by the end of this project.

The second is that the library in this project contains a new _trait_,
`ThreadPool`. It contains the following methods:

- `ThreadPool::new(threads: u32) -> Result<ThreadPool>`

  Creates a new thread pool, immediately spawning the specified number of
  threads.

  Returns an error if any thread fails to spawn. All previously-spawned threads
  are terminated.

- `ThreadPool::spawn<F>(&self, job: F) where F: FnOnce() + Send + 'static`

  Spawn a function into the threadpool.

  Spawning always succeeds, but if the function panics the threadpool continues
  to operate with the same number of threads &mdash; the thread count is not
  reduced nor is the thread pool destroyed, corrupted or invalidated.

By the end of this project there will be several implementations of this
trait, and you will again perform benchmarking to compare them.

This project should not require any changes at all to the client code.

## Project setup

> Create a new cargo project and copy `tests/tests.rs` into it. This project

Reviewer: ```suggestion
Create a new cargo project and copy the `tests` directory into it. This project
```
We may have multiple test files now...

specifying the number of threads in the pool. In this implementation it will go
unused.

_Go ahead and implement this version of `ThreadPool` now_, then we'll
integrate it into the new `KvStore`.

**Test cases to complete**:

  - `naive_thread_pool_*`

## Part 2: Creating a shared `KvEngine`

Before we can integrate the `NaiveThreadPool` into `KvServer` we have to make
the `KvEngine` trait and the `KvStore` implementation (for now you can ignore
the `SledKvsEngine` from the previous project, but you can optionally
re-implement it as an extension to this project).

Recall from the project spec that, this time, our `KvsEngine` takes `self` as
`&self` instead of `&mut self` as previously, and it also implements `Clone`.
More concretely, it looks like

```rust
pub trait KvsEngine: Clone {
    fn set(&self, key: String, value: String) -> Result<()>;

    fn get(&self, key: String) -> Result<Option<String>>;

    fn remove(&self, key: String) -> Result<()>;
}
```

This gives us a lot of clues about the implementation strategy we're pursuing.
First, think about why the engine needs to implement `Clone` when we have a

> multi-threaded implementation. Instead of being coy, this once I'll just tell

Reviewer: I believe "coy" is a new word for most Chinese people... Use "shy" instead?

Author: Great suggestion.

you that each thread is going to need access to the engine. Now think about why
that makes us use `&self` instead of `&mut self`. What do you know about shared
mutable state? By the end of this project be sure you understand the
implications here &mdash; _this is what Rust is all about_.

In this model, `KvsEngine` behaves like a _handle_ to another object, and
because that object is shared between threads, it probably needs to live on the
[heap], and because that shared state can't be mutable it needs to be protected by
some synchronization primitive.

[heap]: https://stackoverflow.com/questions/79923/what-and-where-are-the-stack-and-heap

> So, _hoist the date inside your implementation of `KvEngine`, `KvsStore` onto

Reviewer: I don't understand "hoist the date"... Is it a slang?

Author: it should have said `data`, but "hoist" is slang. It means "pull up" or "pull out": "pull up your data from the stack to the heap". I'll change it.

Author: more precisely, "hoist" isn't _slang_ but using it in this way is jargony.

> the heap usinga thread-safe shared pointer type and protect it behind a lock of

Reviewer: ```suggestion
the heap using a thread-safe shared pointer type and protect it behind a lock of
```

your choosing_.

Remember that because `KvsEngine` requires `Clone` you also need to implement
`Clone` for `KvsStore`.

At this point, if you've carried over your code from previous projects, your
single-threaded `KvsServer` should work once again, but now with a `KvsEngine`

> that can later be shared across thrteads.

Reviewer 2: that can later be shared across thrteads.
                                                      ^^^^^^^^-> Maybe threads?

Author: 👍

abstract the thread pool and compare the performance of yours with an existing.

So, what is a thread pool?

It's nothing complicated. Instead of creating a new thread for every
multi-threaded job to be performed, a thread pool maintains a "pool" of threads,
and reuses those threads instead of creating a new one.

But why?

It's entirely about performance. It saves a small amount of performance, but
when writing high-performance applications, every bit counts. Imagine what it
takes to make a new thread:

You've got to have a call stack for that thread to run on. That's an allocation.
Allocations are pretty cheap, but not as cheap as not allocation. How that call
stack is allocated depends on details of the operating system and runtime, but
can involve locks and syscalls. Syscalls again are not _that_ expensive, but
they are expensive when we're dealing with Rust levels of performance. Reducing
syscalls is a common source of easy optimization. That stack then has to be
carefully initialized so that first [stack _frame_] contains the appropriate
values for the return pointer and other things. In Rust the stack needs to be

> configured with a [guard page]. That's two more syscalls, [to `mmap` and to

Reviewer: Seems that these two syscalls are saved in Linux. Add "except in Linux" for strictness?

Author: :eyeroll: yeah, I saw that after I wrote this and read the std src. It's a bummer because I think this is an illustrative point.

It's also unfortunate that all the dirty mechanics of launching the thread are hidden in libpthread.

I'll do as you suggest.

`mprotect`][mp].

[2mb]: https://github.com/rust-lang/rust/blob/6635fbed4ca8c65822f99e994735bd1877fb063e/src/libstd/sys/unix/thread.rs#L12
[mp]: https://github.com/rust-lang/rust/blob/6635fbed4ca8c65822f99e994735bd1877fb063e/src/libstd/sys/unix/thread.rs#L315

TODO: illustration?

That's just setting up the callstack. It's at least another syscall to create
the new thread, at which point the kernel must do its own accounting for the new
thread.

In Rust libpthread handles most of this complexity.

Then at some point the OS performs a [context switch] onto the new thread. With
a thread pool, all that setup overhead is only done for a few threads, and

> subsequent jobs are simply context switches into that pool of threads.

Reviewer: I think the cost of context switch and scheduling among a large number of threads should be mentioned too.

Author: Can you say more about what you are thinking? I can imagine explaining that scheduling strategies themselves have a cost that has to be taken into consideration.

So how do you build a thread pool?

There are many strategies and tradeoffs, but for this exercise you are going to
write a "round-robin" thread pool. "Round-robin" just means that you have a
circular list of threads and distribute work to them in order. This is the very
simplest work scheduling strategy, but it can be effective. What are the
downsides?

There is one wrinkle in this easy strategy: your thread pool will need to deal
with the case where the spawned function panics &mdash; simply letting panics
destroy the threads in your pool would quickly deplete its pool of threads. So
if a thread in your pool panics you need to make sure that the total number of
threads doesn't decrease. So what should you do? You have two options: let the
thread die and spawn another, or catch the panic and keep the existing thread
running. What are the tradeoffs? You've got to pick one, but leave a comment in
your code explaining your choice.

_Create the `RoundRobinThreadPool` type, implementing `ThreadPool`_.

> The tools at your disposal are `[thread::spawn]`, [`catch_unwind`], [`mpsc`]

Reviewer: It's OK to use `std::mpsc` to implement a round-robin thread pool. But I just realize that it is extremely easy to use mpmc crossbeam channel to implement a thread pool. Every thread just loops and receives tasks from the channel...

Author: That would though not be round-robin, and I'm hoping that round-robin performs particularly badly in comparison to a production threadpool.

I would like to introduce crossbeam though, and it is much easier to write a threadpool with a single queue. Hm.

I'll change it and hope that we can find maybe a work-stealing threadpool that performs better. I have only heard bad things about every rust thread pool... but I'm planning to use rayon and hope for the best.

Author: Maybe I'll try to leave the strategy open, have some background reading about scheduling strategies, and point to a few tools.

channels, and `thread`s [`JoinHandle`]. Depending on your approach you may
use some or all of these.

**Test cases to complete**:

  - `round_robin_thread_pool_*`

Replace the `NaiveThreadPool` used by `KvsServer` with `RoundRobinThreadPool`.
Again your `kvs-server` should still work the same as previously, now with a
slightly more clever multi-threading model. This time you'll want to call
the thread pool constructor with an appropriate number of threads. For

> now you can create a thread per CPU, using the [`num_threads`] crate. We'll

Reviewer: ```suggestion
now you can create a thread per CPU, using the [`num_cpus`] crate. We'll
```

Author: doh

revisit that 

> ## Part 3: Abstract thread pools

Reviewer 2: Maybe it's part 5.

Author: :+1

As in the previous project where you created a `KvsEngine` abstraction to compare
different implementations, now you are going to use the `ThreadPool` abstraction
to do the same.

> If you haven't already, add a second type parameter to `KvServer` to represent

Reviewer: ```suggestion
If you haven't already, add a second type parameter to `KvsServer` to represent
```

affected? What about throughput?

The first step is to write a `ThreadPool` implementation for this naive
approach. Call it `NaiveThreadPool` (it's not _really_ even a thread _pool_
since the threads are not reused, but it needs to conform to our trait for later
comparisons).

We aren't focusing on a more sophisticated implementation now because simply
integrating this solution into our existing design is going to take some effort.
So note that the `ThreadPool::new` constructor takes a `threads` argument
specifying the number of threads in the pool. In this implementation it will go
unused.

_Go ahead and implement this version of `ThreadPool` now_, then we'll
integrate it into the new `KvStore`.

**Test cases to complete**:

  - `naive_thread_pool_*`

## Part 2: Creating a shared `KvEngine`

Before we can integrate the `NaiveThreadPool` into `KvServer` we have to make
the `KvEngine` trait and the `KvStore` implementation (for now you can ignore
the `SledKvsEngine` from the previous project, but you can optionally
re-implement it as an extension to this project).

Recall from the project spec that, this time, our `KvsEngine` takes `self` as

> `&self` instead of `&mut self` as previously, and it also implements `Clone`.

Reviewer: Should also implement `Send`

More concretely, it looks like

```rust
pub trait KvsEngine: Clone {
    fn set(&self, key: String, value: String) -> Result<()>;

    fn get(&self, key: String) -> Result<Option<String>>;

    fn remove(&self, key: String) -> Result<()>;
}
```

This gives us a lot of clues about the implementation strategy we're pursuing.
First, think about why the engine needs to implement `Clone` when we have a
multi-threaded implementation. This once I'll just tell you that each thread is
going to need access to the engine. Now think about why that makes us use
`&self` instead of `&mut self`. What do you know about shared mutable state? By
the end of this project be sure you understand the implications here &mdash;
_this is what Rust is all about_.

In this model, `KvsEngine` behaves like a _handle_ to another object, and
because that object is shared between threads, it probably needs to live on the
[heap], and because that shared state can't be mutable it needs to be protected by
some synchronization primitive.

[heap]: https://stackoverflow.com/questions/79923/what-and-where-are-the-stack-and-heap

> So, _hoist the data inside your implementation of `KvEngine`, `KvsStore` onto

Reviewer: ```suggestion
So, _hoist the data inside your implementation of `KvsEngine`, `KvStore` onto
```

the stack needs to be configured with a [guard page] to prevent stack overflows,
preserving memory safety. That takes two more syscalls, [to `mmap` and to
`mprotect`][mp] (though on Linux in particular, those two syscalls are avoided).

[2mb]: https://github.com/rust-lang/rust/blob/6635fbed4ca8c65822f99e994735bd1877fb063e/src/libstd/sys/unix/thread.rs#L12
[mp]: https://github.com/rust-lang/rust/blob/6635fbed4ca8c65822f99e994735bd1877fb063e/src/libstd/sys/unix/thread.rs#L315

TODO: illustration?

That's just setting up the callstack. It's at least another syscall to create
the new thread, at which point the kernel must do its own internal accounting
for the new thread.

In Rust, the C [libpthread] library handles most of this complexity.

Then at some point the OS performs a [context switch] onto the new thread. When
the thread terminates all that work needs to be undone again.

With a thread pool, all that setup overhead is only done for a few threads, and
subsequent jobs are simply context switches into existing threads in the pool.

### So how do you build a thread pool?

There are many strategies and tradeoffs, but for this exercise you are going to

> use a single shared queue to distribute work to idle threads. That mans that

Reviewer: ```suggestion
use a single shared queue to distribute work to idle threads. That means that
```

  queue, and there is going to be one sender ("producer"), the thread listening
  for TCP connections, and many recievers ("consumers"), the threads in the pool.

2) _how to deal with panicking jobs_ &mdash; your pool runs arbitrary work items.
  If a thread panics, the thread pool needs to recover in some way.

3) _how to deal with shutdown_ &mdash; when the `ThreadPool` object goes out of
  scope it needs to shut down every thread. It must not leave them idle.

These concerns are all intertwined since dealing with each of them may involve
communication and synchronization between threads. Some solutions will be
simple, the solutions to each of these working together gracefully; some
solutions will be complex, the solutions being independent and convoluted.
Choose your data structures carefully and use their capabilities wisely.

You will distribute work by sending messages over some concurrent queue type (a
concurrent queue in Rust typically being a data structure with two connected
types: sender types, and reciever types; and that can send between the two types
any type that implements `Send` + `'static`).

> Messages is Rust are typically represented as enums, with variants for each

Reviewer: ```suggestion
Messages in Rust are typically represented as enums, with variants for each
```

possible message that can be sent, like:

```
enum ThreadPoolMessage {
    RunJob(Box<FnOnce + Send + 'static>),
	Shutdown,
}
```

This tends to be a simpler and more efficient solution than trying to "juggle"
multiple channels for different purposes. Of course, if there is only one type
of message, an enum is not necessary. Now, the above example may or may not be
the full set of messages you need to manage your thread pool, depending on the
design. In particular, shutting down can often be done implicitly if your queue
returns a result indicating that the sender has been destroyed.

There are many types of multi-threaded queues. In Rust the most common is the
[`mpsc`] channel, because it lives in Rust's standard library. This is a
multi-producer, single consumer queue, so using it for your single-queue thread
pool will require a lock of some kind. What's the downside of using a lock here?
There are many other concurrent queue types in Rust, and each has pros and cons.
If you are willing to take a lock on both producer and consumer sides, then you
could even use a `Mutex<VecDeque>`, but there's probably no reason to do that in
production when better solutions exist.

> _Historical note: the existance of channels in Rust's standard library is a bit

Reviewer: ```suggestion
_Historical note: the existence of channels in Rust's standard library is a bit
```

The cargo project, `kvs`, builds a command-line key-value store client called
`kvs-client`, and a key-value store server called `kvs-server`, both of which in
turn call into a library called `kvs`. The client speaks to the server over
a custom protocol.

The interface to the CLI and the library is the same as in the [previous project].
Refer to it. The difference this time is in the parallel implementation, which
will be described as we work through it.

## Project setup

Create a new cargo project and copy `tests/tests.rs` into it. This project
should contain a library named `kvs`, and two executables, `kvs-server` and
`kvs-client`. As with previous projects, add enough definitions that the
test suite builds.

## Part 1: Creating a thread pool

- discuss thread pool designs
- specify a simple interface
- pass test cases

## Part 2: Accepting connections on separate threads

- accept and respond to requests in the thread pool
- this will require sharing the index as well

## Part 3: Abstract thread pools

- like with the sled comparison, abstract the thread
  pool, and add an existing one
- can we do a different _kind_ of abstraction than we did for sled? maybe lets

>   try compile time abstraction and cargo features

Reviewer: Yes, we can have a try. But I don't know what it'll exactly look like yet. Maybe we can add a "xxx-threadpool" features to the project and if we set the feature on through cargo command line arguments, we use the third-party threadpool instead of ours. However, I don't know if it is good or usual to switch implementations through feature flags.

Author: Yeah, it feel unnatural to me to use compile time for this. I'm disinclined to do it. I mean, probably one wouldn't make the threadpool pluggable at all.

## Part 4: Comparing thread pools

- read-centric workload, measuring network throughput
- is just writing more criterion benchmarks useful?

> - can we come up with other tools to measure with,

Reviewer: The students use the client he wrote to benchmark the server. It can be a problem if the bottleneck is the client, for example, students use only one block client to test the server. So it is good if we can use a mature pressure testing tool to test. However, we use a custom protocol but most tools aim at HTTP. Thus, I think we should at least hint students that they should pay attention to concurrency at the client side if they are doing benchmarks.

Reviewer: I have no idea about what else we can benchmark on about a thread pool now

Author: Yes this is a problem I've encountered writing the text. The overhead of using the client is probably going to cause huge problems for benchmarking.

One thing I think we can benchmark is to once again test sled in a multi-threaded setting. It should perform markedly better than our engine.
